### PR TITLE
Add support for singularity installation

### PR DESCRIPTION
**Pull request by jaimesouza**
**Please provide description of the purpose of this pull request, as well as its
motivation and context.**

Create a Singularity class responsible for installing singularity on
slurmd, using .deb (on Ubuntu) and .rpm (on Centos) packages retrieved
from GitHub Releases.

The .deb and .rpm files are slurmd charm resources.

The Singularity install method receives the path to the file from
slurm-charms/charm-slurmd/src/charm.py.

**How was the code tested?**

It works with a slurm-charms up to date version which supports singularity-install action.

Final checklist
- [x] I am the author of these changes, or I have the rights to submit them.
- [ ] I added the relevant changed to the README and/or documentation.
- [x] I self reviewed my own code.
- [ ] I updated the `CHANGELOG` according to the [contributing
  guide](https://omnivector-solutions.github.io/osd-documentation/master/contributing.html#changelog)

### CODE REVIEW

File: slurm_ops_manager/singularity.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        """Return singularity resource name."""
+        return self._stored.resource_name
+
+    def install(self, resource_path: Path):
+        """Install Singularity."""
+
+        logger.debug(f"#### Installing singularity from: {resource_path}")
+
+        cmds = {"ubuntu": f"apt install {resource_path} -y",
+                "centos": f"yum localinstall {resource_path} -y"}
```
Comment (heitorPB) [inline slurm_ops_manager/singularity.py:54]: It is not recommended to run `apt` non-interactively, use `apt-get` instead.

```suggestion
        cmds = {"ubuntu": f"apt-get install --yes {resource_path}",
                "centos": f"yum localinstall --assumeyes {resource_path}"}
```

**Review by heitorPB — CHANGES_REQUESTED** [review]
Nice work! I have some comments for you in the code.

File: slurm_ops_manager/singularity.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        elif self._operating_system == "centos":
+            self._stored.resource_name = "singularity-rpm"
+        else:
+            logger.error(f"#### Unsupported OS: {self._operating_system}")
+            self._stored.resource_name = ""
+
+    @property
+    def installed(self) -> bool:
+        """Return wether singularity is installed."""
+        return self._stored.singularity_installed
```
[the remark this answers is not in the document]
Comment (jaimesouza) [reply slurm_ops_manager/singularity.py:41]: I thought I was going to use it, but I did not. I gonna remove it.

**Review by heitorPB — APPROVED** [review]
Nice work!

File: slurm_ops_manager/slurm_ops.py [hunk 3]
```diff
@@ -46,6 +47,7 @@ def __init__(self, charm, component):
         self.infiniband = Infiniband(charm, component)
         if self._slurm_component == "slurmd":
             self.nvidia = NvidiaGPU(charm, component)
+            self.singularity = Singularity(charm, component)
```
Comment (jamesbeedy) [inline slurm_ops_manager/slurm_ops.py:50]: I'm wondering if we even need to pass the charm and component into the Singularity object. Possibly we change the object to not accept any required initialization parameters.

File: slurm_ops_manager/singularity.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+from pathlib import Path
+
+from ops.framework import Object, StoredState
+from slurm_ops_manager.utils import operating_system
+
+
+logger = logging.getLogger()
+
+
+class Singularity(Object):
```
Comment (jamesbeedy) [inline slurm_ops_manager/singularity.py:15]: Possibly this doesn't need to inherit `Object`....

File: slurm_ops_manager/singularity.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    """Responsible for Singularity operations."""
+
+    _stored = StoredState()
```
Comment (jamesbeedy) [inline slurm_ops_manager/singularity.py:18]: Although, possibly you need to inherit `Object` to use `StoredState()` .... I'm not sure actually

**Comment by jamesbeedy** [comment]
This looks great- though I have one concern.

Does the `Singularity` class need to inherit from `Object` to use `StoredState`? If not, the inheritance should be removed and the signature of __init__ should be changed.

File: slurm_ops_manager/singularity.py (hunk 5 above)
Comment (heitorPB) [reply slurm_ops_manager/singularity.py:18]: Per the ops examples, that is required.

File: slurm_ops_manager/slurm_ops.py (hunk 3 above)
Comment (heitorPB) [reply slurm_ops_manager/slurm_ops.py:50]: That is needed when the class inherits from `Object`

Comment (jamesbeedy) [reply slurm_ops_manager/slurm_ops.py:50]: Do we even need to make it a class then?

Comment (jamesbeedy) [reply slurm_ops_manager/slurm_ops.py:50]: If the class consists of a single install function, possibly we can just define and call a single function?

Comment (heitorPB) [reply slurm_ops_manager/slurm_ops.py:50]: Having a class for that keeps it consistent with having a class for Infiniband and one for Nvidia GPUs. It also makes it easy to add other methods if we need to tweak the settings for Singularity.

We don't _have to_ have a class for it, but it does have some benefits.

**Comment by jamesbeedy** [comment]
+1 nice work!

**Review by jamesbeedy — APPROVED** [review]
+1

**Pull request merged**